hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3480 @@
+    return SYSTEM_ERROR_NONE;
+}
+
+bool BleObject::Broadcaster::advertising() const {
+    return isAdvertising_;
+}
+
+int BleObject::Broadcaster::setAdvertisingParams(const hal_ble_adv_params_t* params) {
+    if (params == nullptr) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
     }
-    // Set MTU size
-    ret = nrf_ble_gatt_att_mtu_periph_set(&g_gatt, DEFAULT_MTU);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "nrf_ble_gatt_att_mtu_periph_set() failed: %u", (unsigned)ret);
-        return halError(ret);
+    CHECK(suspend());
+    if (connHandle_ != BLE_INVALID_CONN_HANDLE) {
+        hal_ble_adv_params_t connectedAdvParams = *params;
+        connectedAdvParams.type = BLE_ADV_SCANABLE_UNDIRECTED_EVT;
+        CHECK(configure(&connectedAdvParams));
+        advParams_.size = sizeof(hal_ble_adv_params_t);
+        memcpy(&advParams_, params, advParams_.size);
+        advParams_.version = BLE_API_VERSION;
+        connectedAdvParams_ = true; // Set the flag after the advParams_ being updated.
+    } else {
+        CHECK(configure(params));
+        advParams_.size = sizeof(hal_ble_adv_params_t);
+        memcpy(&advParams_, params, advParams_.size);
+        advParams_.version = BLE_API_VERSION;
     }
-    return 0;
+    return resume();
 }
 
-int initGap(const char* deviceName) {
-    ble_gap_conn_sec_mode_t secMode = {};
-    // Require no protection
-    BLE_GAP_CONN_SEC_MODE_SET_OPEN(&secMode);
-    // Set device name
-    uint32_t ret = sd_ble_gap_device_name_set(&secMode, (const uint8_t*)deviceName, strlen(deviceName));
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "sd_ble_gap_device_name_set() failed: %u", (unsigned)ret);
-        return halError(ret);
-    }
-    // Set PPCP (Peripheral Preferred Connection Parameters)
-    ble_gap_conn_params_t connParam = {};
-    connParam.min_conn_interval = MIN_CONN_INTERVAL;
-    connParam.max_conn_interval = MAX_CONN_INTERVAL;
-    connParam.conn_sup_timeout = CONN_SUP_TIMEOUT;
-    connParam.slave_latency = SLAVE_LATENCY;
-    ret = sd_ble_gap_ppcp_set(&connParam);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "sd_ble_gap_ppcp_set() failed: %u", (unsigned)ret);
-        return halError(ret);
+int BleObject::Broadcaster::getAdvertisingParams(hal_ble_adv_params_t* params) const {
+    if (params == nullptr) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
     }
-    return 0;
+    memcpy(params, &advParams_, params->size);
+    return SYSTEM_ERROR_NONE;
 }
 
-} // ::
-
-int ble_init(void* reserved) {
-    if (!nrf_sdh_is_enabled()) {
-        LOG(ERROR, "SoftDevice is not enabled");
-        return BLE_ERROR_INVALID_STATE;
-    }
-    // Configure the stack using the SDK's default settings
-    uint32_t ramStart = 0; // Start address of the application RAM
-    ret_code_t ret = nrf_sdh_ble_default_cfg_set(CONN_CFG_TAG, &ramStart);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "nrf_sd
```

Comment:
There will log an error in case of failure. But adding it to cache failed shouldn't affect the following operations. The connected event should be dispatched still to unlock the `connect()` caller.